src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -202,13 +202,281 @@ func (r *downloadRequest) respondFromLocalFile(w http.ResponseWriter, absBasePat
 	if bytesResponded, err := io.Copy(w, file); err != nil {
 		r.Logger.WithError(err).Warn("Failed to copy from cache")
 		if bytesResponded == 0 {
-			return &util.JSONResponse{
-				Code: 500,
-				JSON: jsonerror.NotFound(fmt.Sprintf("Failed to respond with file with media ID %q", r.MediaMetadata.MediaID)),
-			}
+			resErr := jsonerror.InternalServerError()
+			return &resErr
 		}
 		// If we have written any data then we have already responded with 200 OK and all we can do is close the connection
 		return nil
 	}
 	return nil
 }
+
+// getRemoteFile fetches the remote file and caches it locally
+// A hash map of active remote requests to a struct containing a sync.Cond is used to only download remote files once,
+// regardless of how many download requests are received.
+// Note: The named errorResponse return variable is used in a deferred broadcast of the metadata and error response to waiting goroutines.
+// Returns a util.JSONResponse error in case of error
+func (r *downloadRequest) getRemoteFile(cfg *config.MediaAPI, db *storage.Database, activeRemoteRequests *types.ActiveRemoteRequests) (errorResponse *util.JSONResponse) {
+	// Note: getMediaMetadataFromActiveRequest uses mutexes and conditions from activeRemoteRequests
+	mediaMetadata, resErr := r.getMediaMetadataFromActiveRequest(activeRemoteRequests)
+	if resErr != nil {
+		return resErr
+	} else if mediaMetadata != nil {
+		// If we got metadata from an active request, we can respond from the local file
+		r.MediaMetadata = mediaMetadata
+	} else {
+		// Note: This is an active request that MUST broadcastMediaMetadata to wake up waiting goroutines!
+		// Note: errorResponse is the named return variable
+		// Note: broadcastMediaMetadata uses mutexes and conditions from activeRemoteRequests
+		defer r.broadcastMediaMetadata(activeRemoteRequests, errorResponse)
```

Comment:
You are safe to assume that the errorResponse parameter will always be set by the return stmt so that part is fine.
However it wouldn't be set if something panic'd. So this might need to be written as:

```go
defer func() {
     if err := recover(); err != nil {
          resErr := jsonerror.InternalServerError()
          r.broadcastMediaMetadata(activeRemoteRequests, &resErr)
          panic(err)
     }
     r.broadcastMediaMetadata(activeRemoteRequests, errorResponse)
}
```